web_monitoring/html_diff_render.py

```diff
@@ -0,0 +1,209 @@
+    title_meta.attrs['name'] = 'wm-diff-title'
+    soup_new.head.append(title_meta)
+
+    old_head = soup_new.new_tag('template', id='wm-diff-old-head')
+    if soup_old.head:
+        for node in soup_old.head.contents.copy():
+            old_head.append(node)
+    soup_new.head.append(old_head)
+
+    change_styles = soup_new.new_tag(
+        "style",
+        type="text/css",
+        id='wm-diff-style')
+    change_styles.string = """
+        ins {text-decoration: none; background-color: #d4fcbc;}
+        del {text-decoration: none; background-color: #fbb6c2;}"""
+    soup_new.head.append(change_styles)
+
+    # The method we use above to append HTML strings (the diffs) to the soup
+    # results in a non-navigable soup. So we serialize and re-parse :(
+    # (Note we use no formatter for this because proper encoding escape the
+    # tags our differ generated.)
+    soup_new = BeautifulSoup(soup_new.prettify(formatter=None), 'lxml')
+    replacements_new.update(replacements_old)
+    soup_new = _add_undiffable_content(soup_new, replacements_new)
+
+    return soup_new.prettify(formatter='minimal')
+
+
+def _remove_undiffable_content(soup, prefix=''):
+    """
+    Find nodes that cannot be diffed (e.g. <script>, <style>) and replace them
+    with an empty node that has the attribute `wm-diff-replacement="some ID"`
+
+    Returns a tuple of the cleaned-up soup and a dict of replacements.
+    """
+    replacements = {}
+
+    # NOTE: we may want to consider treating <object> and <canvas> similarly.
+    # (They are "transparent" -- containing DOM, but only as a fallback.)
+    for index, element in enumerate(soup.find_all(['script', 'style'])):
+        replacement_id = f'{prefix}-{index}'
+        replacements[replacement_id] = element
+        replacement = soup.new_tag(element.name, **{
+            'wm-diff-replacement': replacement_id
+        })
+        # The replacement has to have text if we want to ensure both old and
+        # new versions of a script are included. Use a single word (so it
+        # can't be broken up) that is unlikely to appear in text.
+        replacement.append(f'$[{replacement_id}]$')
+        element.replace_with(replacement)
+
+    return (soup, replacements)
+
+
+def _add_undiffable_content(soup, replacements):
+    """
+    This is the opposite operation of `_remove_undiffable_content()`. It
+    takes a soup and a replacement dict and replaces nodes in the soup that
+    have the attribute `wm-diff-replacement"some ID"` with the original content
+    from the replacements dict.
+    """
+    for element in soup.select('[wm-diff-replacement]'):
+        replacement = replacements[element['wm-diff-replacement']]
+        if replacement:
+            element.replace_with(replacement)
+
+    return soup
+
+
+def _get_title(soup):
+    return soup.title and soup.title.string or ''
+
+
+def _diff_title(old, new):
+    return ''.join(map(
+        lambda change: ((change[0] == -1 and '<del>{}</del>')
```

Comment:
This is a clever way to implement a switch. I like it, but it took me a minute to parse it. A docstring and/or comment to help future readers might be a good idea.